### PR TITLE
Event branch

### PR DESCRIPTION
**Pull request by HillaryOkello**
In this branch, I worked on:
* Creating and migrating the user and event models.
* Creating the user and event controllers.
* Creating the through table registration to connect event and user tables.
* Creating the views for users and events

### CODE REVIEW

**Comment by mimipeshy** [comment]
### STATUS: CHANGES REQUIRED 🔴 ♻️ 
Hey team @HillaryOkello and @helman101 👋🏾 👋🏾 , you have done a great job with this project, kudos for that 👏🏾 👏🏾 , However, before marking the project as complete, a few changes are needed

**Behavior**

- [ ]  Navigation links are present at the top of the page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/navigation_links_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [ ] The event's show page displays the list of attendants  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/events_attendees_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [ ] The user's show page displays events the user has created and events that the user is attending to separated in past and upcoming  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [ ] Add a sign-in option if a user is already created in the database

#### Models

##### Event

* [ ] it has creation date as attribute ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/event_date_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [ ] scope for upcoming events ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/scopes_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

```ruby
scope :upcoming, -> { where('date >= ?', Time.zone.now) }
```

* [ ] scope for previous events  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/scopes_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

```ruby
scope :previous, -> { where('date < ?',  Time.zone.now) }
```

#### Associations

##### User as event creator

* [ ] User model: `has_many :events, foreign_key: :creator_id` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
  * Migrations
    * [ ] events table has creator_id as the foreign key ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

##### User as an Event attendee ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/attendee_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

* [ ] Registration model `belongs_to :attendee, class_name: "User"`
* [ ] Registration model `belongs_to :attended_event, class_name: "Event"`
* [ ] User model `has_many :attended_events, through: :registrations`
#### Controllers

##### EventsController

###### `index` action

* [ ] lists past events, i.e. `@past_events = Event.previous_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [ ] lists ucomping events, i.e. `@upcoming_events = Event.upcoming_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

##### UsersController

###### `show` action

* [ ] lists events created by user, i.e. `@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [ ] lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [ ] lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

That's all :smile: kindly submit another review when done.

_Feel free to reach out to me on slack if you have any questions or something is unclear, I will be glad to help_
Regards
Peris Ndanu

File: app/views/events/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<div class="container">
+<h2>Past Events</h2>
+<% @event.each do |event| %>
+  <% if event.date < Time.now %>
+      <h4><%= event.name %></h4>
+      <p><%= event.description %></p>
+      <span><%= event.creator.name %></span>
+      <h5><%= event.date%></h5>
+  <% end %> 
+<% end %>
```
Comment (acushlakoncept) [inline app/views/events/index.html.erb]: In your events controller `#index` method you already have `@previous_events = Event.previous_events` defined, just use `@previous_events.each ....` and loop through all past events

File: app/views/events/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+<div class="container">
+<h2>Upcoming Events</h2>
+<% @event.each do |event| %>
+  <% if event.date > Time.now %>
+      <h4><%= event.name %></h4>
+      <p><%= event.description %></p>
+      <span><%= event.creator.name %></span>
+      <h5><%= event.date%></h5>
+  <% end %> 
+<% end %>
```
Comment (acushlakoncept) [inline app/views/events/index.html.erb]: In your events controller `#index` method you already have `@previous_events = Event.previous_events` defined, just use `@upcoming_events = Event.upcoming_events ....` and loop through all upcoming events

File: app/views/users/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  <% else %>
+  <h3 class = "card-title"> Past Events</h3>
+    <% current_user.events.each do |event| %>
+
+      <% if event.date < Time.now %>
+        <li><%= link_to event.name, event %></li>
+        <p class = "card-text ">Description: <%= event.description %> </p>
+      <% elsif current_user.events.empty? %>
+        <p> No past events found </p>
+      <% end %>
```
Comment (acushlakoncept) [inline app/views/users/show.html.erb:22]: Having `@past_events = current_user.attended_events.previous` in your user `#show` method will save you the hassle of having logic in your views. You will only need to loop through `@past_events`

File: app/views/users/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+     <h3 class = "card-title"> Upcoming Events</h3>
+
+     <% current_user.events.each do |event|%>
+
+      <% if event.date > Time.now %>
+        <li><%= link_to event.name, event %></li>
+        <p class = "card-text ">Description: <%= event.description %> </p>
+      <% elsif current_user.events.empty?  %>
+        <p> No future events found </p>
+      <% end %>
```
Comment (acushlakoncept) [inline app/views/users/show.html.erb:35]: Having `@upcoming_events = current_user.attended_events.upcoming` in your user `#show` method will save you the hassle of having logic in your views. You will only need to loop through `@upcoming_events`

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

Both of you have done an awesome job so far 👌. But there are still some changes that are necessary for this project to be approved.

### Behaviour

- [ ] There should be an event page, where users can click to attend an event

- [ ] On the event show page there is no way to get back to the user show page or event page

- [ ] User should not be able to navigate to signup/login page if he/she is already logged in, I could signup and login while still logged in with another user
- [ ] There's an error when I try to create event without date. 
Screen Shot 2021-02-10 at 7 04 29 PM

### Model
✅   You did well on defining the scope of your event model, but I did not see you use it in your view, even though you called it in your controller. See comments below.

- [ ] Add event name and date validation, I should not be able to create an even without any date as well as other fields.

### Associations
#### Migrations
* [ ] events table should have a creator_id as the foreign key ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
👍  The association statement on your event model is nicely done
Lets have same for user `has_many :events, foreign_key: :creator_id`

### Controllers
#### Event Controller
👍  Great job on listing the previous and upcoming events appropriately using the defined scopes from the model.
[OPTIONAL] Don't you think, after an event has been created successfully, the user should be redirected to that event show page?
#### User Controller

- [ ] Instead of using `User.find(params[:id]`, since you already have `current_user` defined, using `@created_events = current_user.events` would be more convenient. Same goes for `
@past_events = current_user.attended_events.previous` and `
@upcoming_events = current_user.attended_events.upcoming` in that way avoid using the logic in your views. See comments I made on your codebase.

<hr>

- [ ] There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  
See affected files below:
> `events/index.html.erb`
> `events/show.html.erb`
> `layouts/application.html.erb`
> `users/show.html.erb`
--------------------------------------------------
**The Resources below may come in handy**
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

- [ ] Your commit messages should be capitalized. You can watch this [video](https://www.loom.com/share/00771f369b7b47fc968c019ecf0d3ba6) on how to edit your commit messages using git-rebase.

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or comment on this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by Georjane — APPROVED** [review]
## Status: PR Approved ✔️ ✔️
### Hi Team 🙌
Nice use of scoping in the view and you have added validations to the event model. Great job. Your project meets the minimum requirements for it to be approved. It's a thumbs up for me. 👍 You can go ahead and merge this PR. 

Good luck on your next project and Happy Coding 💻

**Pull request merged**